node/actors/bft/src/leader/tests.rs

```diff
@@ -1,39 +1,394 @@
+
+    let current_view_leader = util.view_leader(util.current_replica_view());
+    assert_ne!(current_view_leader, util.own_key().public());
+
+    let replica_prepare = util.new_current_replica_prepare(|_| {});
+    let res = util.dispatch_replica_prepare(replica_prepare.clone());
+    assert_matches!(res, Err(ReplicaPrepareError::NotLeaderInView));
+}
+
+#[tokio::test]
+async fn replica_prepare_already_exists() {
+    let mut util = UTHarness::new_with(2).await;
+
+    let view = ViewNumber(2);
+    util.set_replica_view(view);
+    util.set_leader_view(view);
+    assert_eq!(util.view_leader(view), util.own_key().public());
+
+    let replica_prepare = util.new_current_replica_prepare(|_| {});
+    let _ = util.dispatch_replica_prepare(replica_prepare.clone());
+    let res = util.dispatch_replica_prepare(replica_prepare.clone());
+    assert_matches!(
+    res,
+    Err(ReplicaPrepareError::Exists { existing_message }) => {
+            assert_eq!(existing_message, replica_prepare.cast().unwrap().msg);
+        }
+    );
+}
+
+#[tokio::test]
+async fn replica_prepare_num_received_below_threshold() {
+    let mut util = UTHarness::new_with(2).await;
+
+    let view = ViewNumber(2);
+    util.set_replica_view(view);
+    util.set_leader_view(view);
+    assert_eq!(util.view_leader(view), util.own_key().public());
+
+    let replica_prepare = util.new_current_replica_prepare(|_| {});
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_matches!(
+        res,
+        Err(ReplicaPrepareError::NumReceivedBelowThreshold {
+            num_messages: 1,
+            threshold: 2
+        })
+    );
+}
+
+#[tokio::test]
+async fn replica_prepare_invalid_sig() {
+    let mut util = UTHarness::new().await;
+
+    let mut replica_prepare = util.new_current_replica_prepare(|_| {});
+    replica_prepare.sig = util.rng().gen();
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_matches!(res, Err(ReplicaPrepareError::InvalidSignature(_)));
+}
+
+#[tokio::test]
+async fn replica_prepare_invalid_commit_qc() {
+    let mut util = UTHarness::new().await;
+
+    let junk = util.rng().gen::<CommitQC>();
+    let replica_prepare = util.new_current_replica_prepare(|msg| msg.high_qc = junk);
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_matches!(res, Err(ReplicaPrepareError::InvalidHighQC(..)));
+}
+
+#[tokio::test]
+async fn replica_prepare_high_qc_of_current_view() {
+    let mut util = UTHarness::new().await;
+
+    let view = ViewNumber(1);
+    let qc_view = ViewNumber(1);
+
+    util.set_view(view);
+    let qc = util.new_commit_qc(|msg| {
+        msg.view = qc_view;
+    });
+    let replica_prepare = util.new_current_replica_prepare(|msg| msg.high_qc = qc);
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_matches!(
+        res,
+        Err(ReplicaPrepareError::HighQCOfFutureView{ high_qc_view, current_view }) => {
```

Comment:
Nit: Missing space after `HighQCOfFutureView` (should be `HighQCOfFutureView { high_qc_view`).

Reply:
Fixed.

Q: https://github.com/matter-labs/era-consensus/pull/39#discussion_r1408074659